resources/2024-10-dapr-bindings.md

```diff
@@ -0,0 +1,276 @@
+- Adding Application.Dapr/Bindings to the TypeSpec frontend.
+- Adding the corresponding DaprBinding type to the internal representation (DaprRP)
+- Allowing a new type of Dapr component (binding) to be emitted in the backend Kubernetes cluster.
+
+#### Advantages (of each option considered)
+
+The main advantage of this implementation approach is that it is purely additive and will be non-breaking for existing users.
+
+#### Disadvantages (of each option considered)
+
+The main disadvantage of this approach is that it will lead to some code duplication.
+
+#### Proposed Option
+
+This option is the most straightforward and efficient in terms of development time. Since the Dapr Building Block implementation is not yet complete, it would be better to have a simple implementation that can be improved later, even if it introduces some code duplication. Additionally, future implementations of User-Defined Types may lead to a refactor of the entire Dapr RP.
+
+### API design (if applicable)
+
+Aside from the new type, the API will remain identical to the other Dapr Building Blocks.
+
+**typespec/Applications.Dapr/bindings.tsp**
+```TypeSpec
+model DaprBindingResource
+  is TrackedResourceRequired<DaprBindingProperties, "DaprBindings"> {
+  @doc("Binding name")
+  @key("bindingName")
+  @path
+  @segment("bindings")
+  name: ResourceNameString;
+}
+
+@doc("Dapr binding portable resource properties")
+model DaprBindingProperties {
+  ...EnvironmentScopedResource;
+  ...DaprResourceProperties;
+
+  @doc("A collection of references to resources associated with the binding")
+  resources?: ResourceReference[];
+
+  ...RecipeBaseProperties;
+}
+```
+
+### Implementation Details
+
+As for every Dapr Building Block, the implementation will need to create in the Dapr RP:
+- A versioned/unversioned api type converter
+- A dedicated processor for the Dapr Binding type
+
+This implementation will also have some side-effects outside of the Dapr RP:
+- Updates to the allowed resource type in the Portable Resource Renderer
+- Updates to the `getResourceDataByID` function in the Core RP to take the new type into account 
+- Updates to the `ResourceTypesList` variable in the CLI (cli/clients/managements.go) to be able to list the new type
+
+### Error Handling
+
+No new error handling is required. The error handling will remain the same as for other Dapr Building Blocks.
+
+## Test plan
+
+Unit tests must cover:
+- converter functions
+- processor functions
+
+Functional tests proposed:
+- Using a binding in a sample application (the direction, input/output, of the binding is not important for this test). The external system can be something simple like Redis or a [CRON binding](https://docs.dapr.io/reference/components-reference/supported-bindings/cron/) that doesn't require an external system.
+- Using a binding with secret-store indirection, which will require creating and using a secret store in the binding.
+
+## Security
+
+N/A
```

Comment:
We may need some details here on how resources can/should connect and communicate securely with external resources with, say, using secret stores.